File: 733-flood-fill/733-flood-fill.cpp
class Solution {
private:
    int n;
    int m;
    
    int dir[4][2] = {{-1, 0}, {0, 1}, {0, -1}, {1, 0}};
    
    void dfs(vector<vector<int>>& image, int sr, int sc, int prevColor, int newColor) {
        image[sr][sc] = newColor;
        
        for(int d = 0; d < 4; d++) {
            int nr = sr + dir[d][0];
            int nc = sc + dir[d][1];
            
            if(nr >= 0 and nc >= 0 and nr < n and nc < m and image[nr][nc] == prevColor) {
                dfs(image, nr, nc, prevColor, newColor);
            }
        }
    }
    
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        if(image[sr][sc] == color) return image;
        
        n = image.size();
        m = image[0].size();
        
        dfs(image, sr, sc, image[sr][sc], color);
        
        return image;
    }
};

File: 733-flood-fill/733-flood-fill.java
class Solution {
    private void dfs(int r, int c, int[][] image, int pColor, int nColor) {
        image[r][c] = nColor;
        if(r - 1 >= 0 && image[r - 1][c] == pColor) {
            dfs(r - 1, c, image, pColor, nColor);
        }
        if(r + 1 < image.length && image[r + 1][c] == pColor) {
            dfs(r + 1, c, image, pColor, nColor);
        }
        if(c - 1 >= 0 && image[r][c - 1] == pColor) {
            dfs(r, c - 1, image, pColor, nColor);
        }
        if(c + 1 < image[r].length && image[r][c + 1] == pColor) {
            dfs(r, c + 1, image, pColor, nColor);
        }
    }
    
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        if(color == image[sr][sc]) return image;
        dfs(sr, sc, image, image[sr][sc], color);
        return image;
    }
}